lighthouse-core/report/v2/report-renderer.js

```diff
@@ -25,19 +57,184 @@
   }
 
   /**
-   * @param {!Element} element
+   * @param {DetailsJSON} text
+   * @return {Element}
    */
-  render(element) {
-    const pre = document.createElement('pre');
-    pre.textContent = JSON.stringify(this._report, null, 2);
-    element.appendChild(pre);
+  _renderText(text) {
+    const element = this._createElement('div', 'lighthouse-text');
+    element.textContent = text.text;
+    return element;
+  }
+
+  /**
+   * @param {DetailsJSON} block
+   * @return {Element}
+   */
+  _renderBlock(block) {
+    const element = this._createElement('div', 'lighthouse-block');
+    for (const item of block.items) {
+      element.appendChild(this._renderDetails(item));
+    }
+    return element;
+  }
+
+  /**
+   * @param {DetailsJSON} list
+   * @return {Element}
+   */
+  _renderList(list) {
+    const element = this._createElement('details', 'lighthouse-list');
+    if (list.header) {
+      const summary = this._createElement('summary', 'lighthouse-list__header');
+      summary.textContent = list.header.text;
+      element.appendChild(summary);
+    }
+
+    const items = this._createElement('div', 'lighthouse-list__items');
+    for (const item of list.items) {
+      items.appendChild(this._renderDetails(item));
+    }
+    element.appendChild(items);
+    return element;
+  }
+
+  /**
+   * @param {Error} e
+   * @return {Element}
+   */
+  _renderException(e) {
+    const element = this._createElement('div', 'lighthouse-exception');
+    element.textContent = String(e.stack);
+    return element;
+  }
+
+  /**
+   * @param {ReportJSON} report
+   * @return {Element}
+   */
+  _renderReport(report) {
+    const element = this._createElement('div', 'lighthouse-report');
+    for (const category of report.reportCategories) {
+      element.appendChild(this._renderCategory(category));
+    }
+    return element;
+  }
+
+  /**
+   * @param {CategoryJSON} category
+   * @return {Element}
+   */
+  _renderCategory(category) {
+    const element = this._createElement('div', 'lighthouse-category');
+    element.appendChild(this._renderScore(category.score, category.name, category.description));
+    for (const audit of category.audits) {
+      element.appendChild(this._renderAudit(audit));
+    }
+    return element;
+  }
+
+  /**
+   * @param {AuditJSON} audit
+   * @return {Element}
+   */
+  _renderAudit(audit) {
+    const element = this._createElement('div', 'lighthouse-audit');
+    let title = audit.result.description;
+    if (audit.result.displayValue) {
+      title += ': ' + audit.result.displayValue;
+    }
+    element.appendChild(this._renderScore(audit.score, title, audit.result.helpText));
+    if (audit.result.details) {
+      element.appendChild(this._renderDetails(audit.result.details));
+    }
+    return element;
   }
 };
+})();
+
+/** @typedef {{type: string, text: string|undefined, header: DetailsJSON|undefined, items: Array<DetailsJSON>|undefined}} */
+let DetailsJSON; // eslint-disable-line no-unused-vars
```

Comment:
what will break if we stick these in the iife? needed for closure in other files?